File: 0023 - Merge k Sorted Lists/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeKLists(struct ListNode** lists, int listsSize){
    int n = listsSize;
    struct ListNode *head;
    if(n == 0) head = NULL;
    else if(n == 1) head = lists[0];
    else if(n == 2){
        struct ListNode **t = &head;
        while(lists[0] && lists[1]){
            if(lists[0]->val < lists[1]->val) *t = lists[0], lists[0] = lists[0]->next;
            else *t = lists[1], lists[1] = lists[1]->next;
            t = &((*t)->next);
        }
        *t = lists[0]?lists[0]:lists[1];
    }
    else{
        struct ListNode *s[2];
        s[0] = mergeKLists(lists, n/2);
        s[1] = mergeKLists(lists+n/2, n-n/2);
        head = mergeKLists(s, 2);
    }
    return head;
}

File: 0023 - Merge k Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def lt(self, other):
    return self.val <= other.val
ListNode.__lt__ = lt

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        import heapq
        head = now = ListNode()
        h = [i for i in lists if i]
        heapq.heapify(h)
        while h:
            now.next = heapq.heappop(h)
            now = now.next
            if now.next:
                heapq.heappush(h, now.next)
        return head.next